warehouse/migrations/versions/93388f06f5e7_draft_releases.py

```diff
@@ -0,0 +1,77 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Draft Releases
+
+Revision ID: 93388f06f5e7
+Revises: 80018e46c5a4
+Create Date: 2020-09-29 18:23:35.798571
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+
+revision = "93388f06f5e7"
+down_revision = "80018e46c5a4"
+
+# Note: It is VERY important to ensure that a migration does not lock for a
+#       long period of time and to ensure that each individual migration does
+#       not break compatibility with the *previous* version of the code base.
+#       This is because the migrations will be ran automatically as part of the
+#       deployment process, but while the previous version of the code is still
+#       up and running. Thus backwards incompatible changes must be broken up
+#       over multiple migrations inside of multiple pull requests in order to
+#       phase them in over multiple deploys.
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column("releases", sa.Column("project_name", sa.Text(), nullable=True))
+    op.add_column("releases", sa.Column("published", sa.DateTime(), nullable=True))
+    # ### end Alembic commands ###
+    # Fill the project_name and published columns
+    op.execute(
+        """
+        UPDATE releases AS target SET published = (
+        SELECT created
+        FROM releases AS source WHERE source.id = target.id
```

Comment:
I believe this will be very hard for the DB. If I understand correctly, we're making a single UPDATE statement that will lock the whole releases DB, and will make 2 million SELECT subqueries.

Reply:
I don't know what the habit for migration queries is in warehouse, but I suggest chunking the request: do the `UPDATE` with a `WHERE published IS NULL LIMIT 10000` and loop until it returns 0 rows updated. Maybe with a sleep() between each loop.

Reply:
Also we'll have a problem: what about all the releases created between the migration and the moment where the Warehouse code will be deployed ? I believe they will be created without published date and thus will be considered drafts as soon as we deploy the new code.

Reply:
(Continuing the discussion after a huge hiatus, sorry)
Well, if I understand things correctly, there are 3 phases:
- A. Before we run the migration
- B. After we run the migration, and before we deploy the code
- C. After we deploy the code

My question refers to phase B: we have added the nullable `published` column and run the big UPDATE query, but at this point, the code doesn't know about the `published` column, so any new release, our INSERT query will not fill this column, and it will get the default value which is null.

I think we should default this column to `NOW()`, and only set it to null when we want to do it so explicitely. Once the code is deploy, we may add another migration that drops the default because from now on, we'll always explicitely set a value, either `NULL` or a date, but it's harming no-one to leave `NOW()` as a default.

Reply:
Took your advice, so I'm resolving this convo :)